File: crates/xray-core/tests/golden.rs
//! Golden test: the core reproduces every blessed `expected/<spec>.json`
//! byte-for-byte.
//!
//! For each spec under `golden/specs/`, build an [`Xray`], load the shared
//! `golden/data.json`, ask for the full-window `frame`, and assert the reply
//! string equals the blessed `golden/expected/<spec>.json` exactly. The reply
//! is `serde_json::to_string(&frame)` — the same compact string every binding
//! returns from a `frame` command and the same string the CLI's `--format json`
//! prints — so this file is the in-core anchor of the cross-language
//! byte-equality guarantee. The blessed files carry a trailing newline from the
//! CLI's `println!`; the command reply does not, so the expectation is trimmed.

use std::fs;
use std::path::{Path, PathBuf};

use xray_core::Xray;

/// The repository-root `golden/` directory, resolved from this crate's manifest.
fn golden_dir() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("..")
        .join("..")
        .join("golden")
}

#[test]
fn every_spec_reproduces_its_blessed_frame() {
    let golden = golden_dir();
    let dataset = fs::read_to_string(golden.join("data.json")).unwrap();

    let mut specs: Vec<PathBuf> = fs::read_dir(golden.join("specs"))
        .unwrap()
        .map(|entry| entry.unwrap().path())
        .filter(|path| path.extension().is_some_and(|ext| ext == "json"))
        .collect();
    specs.sort();
    assert!(!specs.is_empty(), "no golden specs found");

    for spec_path in specs {
        let name = spec_path.file_stem().unwrap().to_str().unwrap();
        let spec_json = fs::read_to_string(&spec_path).unwrap();
        let expected = fs::read_to_string(golden.join("expected").join(format!("{name}.json")))
            .unwrap_or_else(|e| panic!("{name}: missing expected: {e}"));

        let mut xray = Xray::new(&spec_json).unwrap();
        let load = format!(r#"{{"cmd":"load","dataset":{dataset}}}"#);
        xray.command_json(&load).unwrap();
        let frame = xray.command_json(r#"{"cmd":"frame"}"#).unwrap();

        assert_eq!(frame, expected.trim_end(), "{name}: golden mismatch");
    }
}

File: bindings/r/tests/golden.R
## Golden-corpus cases for the wickra-xray R binding.
##
## Lives outside run_tests.R and outside the built tarball (see .Rbuildignore).
## The corpus is at the repository root, above this package, so these cases only
## resolve when run from there -- which CI does explicitly. `R CMD check` runs
## everything under tests/ from the built tarball, where the corpus does not
## exist, and that is what r-universe runs on every platform: a shipped test that
## reasons about the repository above it passes for the wrong reason at best and
## fails every platform binary at worst.
##
## Run from the repository root:
##
##     Rscript bindings/r/tests/golden.R

library(wickraxray)

## cross-language golden parity: build the xray from each committed
## golden/specs/*.json, load the shared golden/data.json and read back the frame,
## and assert the response equals golden/expected/<spec>.json byte-for-byte. The
## binding returns the core's compact command output verbatim, so byte equality
## is the exact cross-language parity check.
golden_dir <- function() {
  d <- normalizePath(getwd(), mustWork = FALSE)
  for (i in seq_len(8)) {
    g <- file.path(d, "golden")
    if (dir.exists(file.path(g, "specs"))) {
      return(g)
    }
    d <- dirname(d)
  }
  NULL
}

g <- golden_dir()
if (is.null(g)) {
  stop("golden fixtures not found; run this from the repository root")
}

dataset <- trimws(paste(
  readLines(file.path(g, "data.json"), warn = FALSE), collapse = "\n"
))
load_all <- paste0('{"cmd":"load","dataset":', dataset, '}')
specs <- list.files(file.path(g, "specs"), pattern = "\\.json$", full.names = TRUE)
if (length(specs) == 0) {
  stop("golden/specs holds no spec; the parity check would assert nothing")
}
for (spec_path in specs) {
  name <- basename(spec_path)
  spec_json <- paste(readLines(spec_path, warn = FALSE), collapse = "\n")
  expected <- trimws(paste(
    readLines(file.path(g, "expected", name), warn = FALSE), collapse = "\n"
  ))
  gxray <- wkxray_new(spec_json)
  invisible(wkxray_command(gxray, load_all))
  got <- wkxray_command(gxray, '{"cmd":"frame"}')
  stopifnot(identical(trimws(got), expected))
}

cat("wickra-xray R golden parity passed (", length(specs), " specs )\n", sep = "")
